Sources/MongoSwift/BSON/Document.swift

```diff
@@ -255,10 +245,27 @@ public struct Document: ExpressibleByDictionaryLiteral, ExpressibleByArrayLitera
 
     /// Appends the key/value pairs from the provided `doc` to this `Document`. 
     public mutating func merge(_ doc: Document) throws {
+        self.checkIsUniquelyReferenced()
         if !bson_concat(self.data, doc.data) {
             throw MongoError.bsonEncodeError(message: "Failed to merge \(doc) with \(self)")
         }
     }
+
+    /// Checks if the document is uniquely referenced. If not, makes a copy
+    /// of the underlying `bson_t` and lets the copy/copies keep the original.
+    /// This allows us to provide value semantics for `Document`s. 
+    /// This happens if someone copies a document and modifies it.
+    /// For example: 
+    ///  let doc1: Document = ["a": 1]
+    ///  var doc2 = doc1
+    ///  doc2["b"] = 2
+    /// Therefore, this function should be called just before we are about to
+    /// modify a document - either by setting a value or merging in another doc.
+    private mutating func checkIsUniquelyReferenced() {
```

Comment:
nit: this is doing more than just checking if its uniquely referenced, I think it sounds more like `copyStorageIfRequired` or something

Reply:
👍 fixed